lib/utils.js

```diff
@@ -82,411 +82,450 @@ var truncateMessage = function(parts, max) {
-     * that list matches the beginning of the phone then we set up a response
-     * with a denied state. The pending message process will ignore these
-     * messages and those reports will be left without an auto-reply. The
-     * denied messages still show up in the messages export.
-     *
-     * @param {String} from - Recipient phone number
-     * @returns {Boolean}
-     */
-    isOutgoingAllowed: function(from) {
-        var self = module.exports,
-            conf = config.get('outgoing_deny_list') || '';
-        if (!from) {
-            return true;
+        if (results.rows.length > 1) {
+          console.warn('More than one patient person document for shortcode ' + patientShortcodeId);
         }
-        if (self._isMessageFromGateway(from)) {
-            return false;
+
+        return callback(null, results.rows[0].id);
+    });
+  },
+  /*
+   * Used to avoid infinite loops of auto-reply messages between gateway and
+   * itself.
+   */
+  _isMessageFromGateway: from => {
+    const gw = config.get('gateway_number');
+    if (typeof gw === 'string' && typeof from === 'string') {
+      return phoneUtil.isNumberMatch(gw, from) >= 3;
+    }
+    return false;
+  },
+  hydrateDoc: (db, id, callback) => {
+    db.medic.view('medic-client', 'docs_by_id_lineage', {
+      startkey: [ id ],
+      endkey: [ id, {} ],
+      include_docs: true
+    }, (err, lineage) => {
+      if (err) {
+        return callback(err);
+      }
+      const contactIds = lineage.rows
+        .map(row => row.doc && row.doc.contact && row.doc.contact._id)
+        .filter(id => !!id);
+      db.medic.fetch({ keys: contactIds }, (err, contacts) => {
+        if (err) {
+          return callback(err);
         }
-        return _.every(conf.split(','), function(s) {
-            // ignore falsey inputs
-            if (!s) {
-                return true;
-            }
-            // return false if we get a case insensitive starts with match
-            return from.toLowerCase().indexOf(s.trim().toLowerCase()) !== 0;
-        });
-    },
-    /*
-     * Given a patient "shortcode" (as used in SMS reports), return the _id
-     * of the patient's person contact to the caller
-     */
-    getPatientContactUuid: function(db, patientShortcodeId, callback) {
-        db.medic.view('medic', 'patient_by_patient_shortcode_id',
-            {
-                key: patientShortcodeId
-            },
-            function(err, results) {
-                if (err) {
-                    return callback(err);
-                }
-
-                if (!results.rows.length) {
-                    return callback();
-                }
-
-                if (results.rows.length > 1) {
-                    console.warn('More than one patient person document for shortcode ' + patientShortcodeId);
-                }
-
-                return callback(null, results.rows[0].id);
+        contacts.rows.forEach(contactRow => {
```

Comment:
make a `hydrateLineageRows` function, that'll make it much easier to read because it's named. (or use a comment, but you know, comments are documentation)